File: app/backend/routers/stocks.py
"""
/api/stocks - stock_master一覧
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from typing import Optional
import main
from auth import require_proxy
from redis_cache import cache_get as _cache_get, cache_set as _cache_set

router = APIRouter(dependencies=[Depends(require_proxy)])

_STOCKS_TTL = 86400  # 24時間 (stock_master は静的、管理者更新時のみ)


class StockMaster(BaseModel):
    """銘柄マスターのレスポンス"""
    ticker: str
    name: Optional[str]
    sector: Optional[str]
    industry: Optional[str]
    price_category: Optional[str]
    watchlist_category: Optional[str]
    market_cap: Optional[int]
    exchange: Optional[str]
    is_active: bool


class StocksResponse(BaseModel):
    """銘柄一覧レスポンス"""
    stocks: list[StockMaster]
    total: int


@router.get("", response_model=StocksResponse)
async def get_stocks(
    category: Optional[str] = Query(None, description="価格カテゴリでフィルタ: penny, mid, large"),
    watchlist: Optional[str] = Query(None, description="ウォッチリストカテゴリでフィルタ"),
    active_only: bool = Query(True, description="アクティブのみ取得"),
):
    """
    stock_master一覧を取得

    - **category**: penny (<$20), mid ($20-100), large ($100+)
    - **watchlist**: robotics, defense, ai など
    - **active_only**: is_active=trueのみ
    """
    cache_key = f"stocks:master:{category or 'all'}:{watchlist or 'all'}:{active_only}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return StocksResponse(**cached)

    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        query = supabase.table("stock_master").select("*")

        if active_only:
            query = query.eq("is_active", True)

        if category:
            query = query.eq("price_category", category)

        if watchlist:
            query = query.eq("watchlist_category", watchlist)

        # ティッカーでソート
        query = query.order("ticker")

        result = query.execute()

        stocks = [StockMaster(**row) for row in result.data]

        response = StocksResponse(stocks=stocks, total=len(stocks))
        _cache_set(cache_key, response.model_dump(), ttl=_STOCKS_TTL)
        return response

    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/{ticker}", response_model=StockMaster)
async def get_stock(ticker: str):
    """
    特定の銘柄情報を取得
    """
    cache_key = f"stocks:master:single:{ticker.upper()}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return StockMaster(**cached)

    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        result = supabase.table("stock_master").select("*").eq("ticker", ticker.upper()).single().execute()

        if not result.data:
            raise HTTPException(status_code=404, detail=f"Stock {ticker} not found")

        stock = StockMaster(**result.data)
        _cache_set(cache_key, stock.model_dump(), ttl=_STOCKS_TTL)
        return stock

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/categories/list")
async def get_categories():
    """
    利用可能なカテゴリ一覧を取得
    """
    cache_key = "stocks:master:categories"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        result = supabase.table("stock_master").select("price_category, watchlist_category").execute()

        price_categories = set()
        watchlist_categories = set()

        for row in result.data:
            if row.get("price_category"):
                price_categories.add(row["price_category"])
            if row.get("watchlist_category"):
                watchlist_categories.add(row["watchlist_category"])

        response = {
            "price_categories": sorted(list(price_categories)),
            "watchlist_categories": sorted(list(watchlist_categories)),
        }
        _cache_set(cache_key, response, ttl=_STOCKS_TTL)
        return response

    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

File: app/worker/src/routes/stocks.ts
/**
 * /api/stocks — 銘柄マスター CRUD (認証不要)
 * ポート元: app/backend/routers/stocks.py
 */

import type { Env } from '../env';
import { getSupabase } from '../lib/supabase';
import { jsonResponse, errorResponse } from '../lib/response';
import { isValidTicker } from '../lib/validation';

export async function handleStocks(
  request: Request,
  env: Env,
  cors: Record<string, string> | null,
): Promise<Response> {
  if (request.method !== 'GET') return errorResponse('Method not allowed', 405, cors);

  const supabase = getSupabase(env);
  const url = new URL(request.url);
  const path = url.pathname.replace('/api/stocks', '') || '/';

  try {
    if (path === '/') return listStocks(url, supabase, cors);
    if (path === '/categories/list') return listCategories(supabase, cors);

    const tickerMatch = path.match(/^\/([^/]+)$/);
    if (tickerMatch) return getStock(supabase, tickerMatch[1], cors);

    return errorResponse('Not Found', 404, cors);
  } catch {
    return errorResponse('Internal server error', 500, cors);
  }
}

// GET /api/stocks
async function listStocks(
  url: URL,
  supabase: ReturnType<typeof getSupabase>,
  cors: Record<string, string> | null,
) {
  let query = supabase.table('stock_master').select('*');

  const category = url.searchParams.get('category');
  if (category) query = query.eq('price_category', category);

  const watchlist = url.searchParams.get('watchlist');
  if (watchlist) query = query.eq('watchlist_category', watchlist);

  const activeOnly = url.searchParams.get('active_only') !== 'false';
  if (activeOnly) query = query.eq('is_active', true);

  const { data } = await query.order('ticker');
  const stocks = data || [];
  return jsonResponse({ stocks, total: stocks.length }, 200, cors);
}

// GET /api/stocks/:ticker
async function getStock(
  supabase: ReturnType<typeof getSupabase>,
  ticker: string,
  cors: Record<string, string> | null,
) {
  const t = ticker.toUpperCase();
  if (!isValidTicker(t)) return errorResponse('Invalid ticker format', 400, cors);

  const { data } = await supabase
    .table('stock_master').select('*').eq('ticker', t).limit(1);
  if (!data?.length) return errorResponse(`Stock ${t} not found`, 404, cors);
  return jsonResponse(data[0], 200, cors);
}

// GET /api/stocks/categories/list
async function listCategories(
  supabase: ReturnType<typeof getSupabase>,
  cors: Record<string, string> | null,
) {
  const { data } = await supabase
    .table('stock_master').select('price_category, watchlist_category');
  const rows = data || [];

  const priceCategories = [...new Set(rows.map(r => r.price_category).filter(Boolean))].sort();
  const watchlistCategories = [...new Set(rows.map(r => r.watchlist_category).filter(Boolean))].sort();

  return jsonResponse({ price_categories: priceCategories, watchlist_categories: watchlistCategories }, 200, cors);
}
